### PR TITLE
Add a HealthChecksUI sample

### PR DESCRIPTION
**Pull request by DamianEdwards**
Builds on preview.4 updates.

This sample demonstrates adding a custom container-based resource that hosts the [AspNetCore.HealthChecksUI](https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks?tab=readme-ov-file#HealthCheckUI).

```csharp
var builder = DistributedApplication.CreateBuilder(args);

var cache = builder.AddRedis("cache");

var apiService = builder.AddProject<Projects.HealthChecksUI_ApiService>("apiservice");

var webFrontend = builder.AddProject<Projects.HealthChecksUI_Web>("webfrontend")
    .WithReference(cache)
    .WithReference(apiService);

builder.AddHealthChecksUI("healthchecksui")
    .WithReference(apiService, endpointName: "http")
    .WithReference(webFrontend);

builder.Build().Run();
```

image

Key detail to note is it defaults to adding an additional endpoint to monitored projects, dedicated to serving the health checks info the UI needs. This is obviously more sensitive than the default health checks payload and is why it defaults to splitting it onto a new endpoint. To configure this it ends up setting the `AllowedHosts` config value on the projects to configure the host filtering middleware, so that the `RequireHost` call on the health checks endpoint is enforced.

In draft for now as this will require updates to `Aspire.Hosting` to support deployment, e.g. currently blocked by dotnet/aspire#2578, but would like to start getting feedback and using this to help drive more advanced scenarios like deployment of projects with multiple endpoints configured, configuring `AllowedHosts` as part of deployment, etc.

### CODE REVIEW

File: samples/HealthChecksUI/HealthChecksUI.AppHost/HealthChecksUIContainerResource.cs [hunk 1]
```diff
@@ -0,0 +1,254 @@
+﻿using Aspire.Hosting.Lifecycle;
```
Comment (DamianEdwards) [inline samples/HealthChecksUI/HealthChecksUI.AppHost/HealthChecksUIContainerResource.cs:1]: This is the file with the bulk of the logic so focus here.

File: samples/HealthChecksUI/HealthChecksUI.AppHost/HealthChecksUIExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    /// <returns></returns>
+    public static IResourceBuilder<HealthChecksUIResource> AddHealthChecksUI(
+        this IDistributedApplicationBuilder builder,
+        string name,
+        int? port = null,
+        string? tag = null)
+    {
+        builder.Services.TryAddLifecycleHook<HealthChecksUILifecycleHook>();
+
+        var keycloakContainer = new HealthChecksUIResource(name);
```
Comment (mitchdenny) [inline samples/HealthChecksUI/HealthChecksUI.AppHost/HealthChecksUIExtensions.cs]: keycloakContainer?

Comment (DamianEdwards) [reply samples/HealthChecksUI/HealthChecksUI.AppHost/HealthChecksUIExtensions.cs]: Already fixed 😁

File: samples/HealthChecksUI/HealthChecksUI.AppHost/HealthChecksUIContainerResource.cs [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+                    {
+                        // Running during dev inner-loop
+                        // Set/update the environment variable for health checks endpoint URL
+                        if (project.Resource.GetAllocatedEndpoint(monitoredProject.EndpointName) is { } allocatedEndpoint
+                            && Uri.TryCreate(allocatedEndpoint.UriString, UriKind.Absolute, out var baseUri))
+                        {
+                            var healthChecksUri = new Uri(baseUri, monitoredProject.ProbePath);
+
+                            context.EnvironmentVariables.AddDelimitedValues(HEALTHCHECKSUI_URLS,
+                                [healthChecksUri.ToString(), healthChecksUri.ToString().Replace("localhost", "host.docker.internal")]);
```
Comment (eerhardt) [inline samples/HealthChecksUI/HealthChecksUI.AppHost/HealthChecksUIContainerResource.cs:146]: What about podman?

File: samples/HealthChecksUI/infra/main.bicep [hunk 4]
```diff
@@ -0,0 +1,42 @@
+targetScope = 'subscription'
```
[the remark this answers is not in the document]
Comment (DamianEdwards) [reply samples/HealthChecksUI/infra/main.bicep:1]: No, this is still a WIP and I added bicep here as part of proving out the scenario. I'll change this to a draft and retarget it to preview.5

File: samples/HealthChecksUI/HealthChecksUI.AppHost/HealthChecksUIContainerResource.cs (hunk 3 above)
Comment (DamianEdwards) [reply samples/HealthChecksUI/HealthChecksUI.AppHost/HealthChecksUIContainerResource.cs:146]: Yeah this came up in another issue. Is it possible to detect the container runtime from here yet?

File: samples/HealthChecksUI/HealthChecksUI.AppHost/HealthChecksUIExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        var monitoredProject = new MonitoredProject(project, endpointName: endpointName, probePath: probePath)
+        {
+            EndpointPort = endpointPort
+        };
+        builder.Resource.MonitoredProjects.Add(monitoredProject);
+
+        return builder;
+    }
+
+    // TODO: Support referencing supported database containers and/or connection strings and configuring the HealthChecksUI container to use them
```
[the remark this answers is not in the document]
Comment (DamianEdwards) [reply samples/HealthChecksUI/HealthChecksUI.AppHost/HealthChecksUIExtensions.cs:58]: Plan is I either extend the sample to do it, or I remove the comment 😃

**Comment by DamianEdwards** [comment]
Going to close for now as I'm using this branch to flesh out what's needed for updates to endpoint modeling for both internal/external and HTTPS support.

**Pull request closed without merging**